plugin/forward/connect.go

```diff
@@ -14,6 +15,19 @@ import (
 	"golang.org/x/net/context"
 )
 
+func (p *Proxy) readTimeout() time.Duration {
+	rtt := time.Duration(atomic.LoadInt64(&p.avgRtt))
+	if rtt < timeout/2 {
+		return 2 * rtt
+	}
+	return timeout
+}
+
+func (p *Proxy) updateRtt(newRtt time.Duration) {
+	rtt := time.Duration(atomic.LoadInt64(&p.avgRtt))
+	atomic.AddInt64(&p.avgRtt, int64((newRtt-rtt)/rttCount))
```

Comment:
The load is atomic and the add is atomic but the two together are not. So you could for example:

thread 1: load
thread 2: load
thread 2: add
thread 1: add

do we care?

Reply:
Yes, I know that combination of these two lines is not atomic. But we don't need precise value here, and to my mind the result looks acceptable.
Let me know if you prefer more safe and precise solution here (e.g. based on mutex)